src/crimson/mon/MonClient.cc

```diff
@@ -0,0 +1,500 @@
+  m->handle = tid;
+  m->what = map;
+  auto& req = version_reqs[tid];
+  return active_con->get_conn()->send(MessageRef{m, false}).then([&req] {
+    return req.get_future();
+  });
+}
+
+seastar::future<>
+Client::handle_get_version_reply(Ref<MMonGetVersionReply> m)
+{
+  if (auto found = version_reqs.find(m->handle);
+      found != version_reqs.end()) {
+    auto& result = found->second;
+    logger().trace("{}: {} returns {}",
+                 __func__, m->handle, m->version);
+    result.set_value(m->version, m->oldest_version);
+    version_reqs.erase(found);
+  } else {
+    logger().warn("{}: version request with handle {} not found",
+                __func__, m->handle);
+  }
+  return seastar::make_ready_future<>();
+}
+
+seastar::future<> Client::handle_mon_command_ack(Ref<MMonCommandAck> m)
+{
+  const auto tid = m->get_tid();
+  if (auto found = mon_commands.find(tid);
+      found != mon_commands.end()) {
+    auto& result = found->second;
+    logger().trace("{} {}", __func__, tid);
+    result.set_value(m->r, m->rs, std::move(m->get_data()));
+    mon_commands.erase(found);
+  } else {
+    logger().warn("{} {} not found", __func__, tid);
+  }
+  return seastar::make_ready_future<>();
+}
+
+seastar::future<> Client::handle_log_ack(Ref<MLogAck> m)
+{
+  // XXX
+  return seastar::make_ready_future<>();
+}
+
+seastar::future<> Client::handle_config(Ref<MConfig> m)
+{
+  return ceph::common::local_conf().set_mon_vals(m->config);
+}
+
+std::vector<unsigned> Client::get_random_mons(unsigned n) const
+{
+  uint16_t min_priority = std::numeric_limits<uint16_t>::max();
+  for (const auto& m : monmap.mon_info) {
+    if (m.second.priority < min_priority) {
+      min_priority = m.second.priority;
+    }
+  }
+  vector<unsigned> ranks;
+  for (const auto& m : monmap.mon_info) {
+    if (m.second.priority == min_priority) {
+      ranks.push_back(monmap.get_rank(m.first));
+    }
+  }
+  std::random_device rd;
+  std::default_random_engine rng{rd()};
+  std::shuffle(ranks.begin(), ranks.end(), rng);
+  if (n == 0 || n > ranks.size()) {
+    return ranks;
+  } else {
+    return {ranks.begin(), ranks.begin() + n};
+  }
+}
+
+seastar::future<> Client::build_initial_map()
+{
+  return monmap.build_initial(ceph::common::local_conf());
+}
+
+seastar::future<> Client::authenticate(std::chrono::seconds seconds)
+{
+  return seastar::with_timeout(
+    seastar::lowres_clock::now() + seconds,
+    reopen_session(-1));
+}
+
+seastar::future<> Client::stop()
+{
+  return tick_gate.close().finally([this] {
+    return timer.cancel();
+  }).then([this] {
+    if (active_con) {
+      return active_con->close();
+    } else {
+      return seastar::make_ready_future<>();
+    }
+  });
+}
+
+seastar::future<> Client::reopen_session(int rank)
+{
+  vector<unsigned> mons;
+  if (rank >= 0) {
+    mons.push_back(rank);
+  } else {
+    // const auto parallel =
+    //   ceph::common::local_conf().get_val<uint64_t>("mon_client_hunt_parallel");
```

Comment:
planning to add this later?

Reply:
oh, i forgot to uncomment this change when pushing this change. hunting-in-parallel indeed works. will uncomment it once #23957 is merged.